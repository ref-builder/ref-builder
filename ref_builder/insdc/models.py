"""Models for data following INSDC standrds."""

import contextlib
from collections.abc import Iterator
from enum import StrEnum
from typing import Annotated, Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    RootModel,
    field_validator,
)

from ref_builder.models import Molecule, MolType


class FTGenbankLexicon(StrEnum):
    """Table keys used in Genbank."""

    FEATURE_KEY = "GBFeature_key"
    FEATURE_INTERVALS = "GBFeature_intervals"
    FEATURE_QUALIFIERS = "GBFeature_quals"
    INTERVAL_FROM = "GBInterval_from"
    INTERVAL_TO = "GBInterval_to"
    QUALIFIER_NAME = "GBQualifier_name"
    QUALIFIER_VALUE = "GBQualifier_value"


class INSDCDatabaseCrossReference(BaseModel):
    """A cross reference qualifier.

    Based on the INSDC guidelines for the /db_xref qualifier

    Reference:
    https://www.insdc.org/submitting-standards/dbxref-qualifier-vocabulary/
    """

    database: str

    identifier: int | str

    @classmethod
    def from_string(cls, db_xref: str) -> "INSDCDatabaseCrossReference":
        """Return labelled data from a standard ``db_xref`` string."""
        database, identifier = db_xref.split(":")[:2]

        return INSDCDatabaseCrossReference.model_validate(
            {"database": database, "identifier": identifier}
        )

    @field_validator("identifier", mode="after")
    @classmethod
    def convert_integer_identifier(cls, v: str | int) -> str | int:
        """Check if identifier can be converted to an integer and convert if so."""
        with contextlib.suppress(ValueError):
            return int(v)

        return v


class INSDCMolType(StrEnum):
    """In vivo molecule types.

    Based on the INSDC controlled vocabulary list for the /mol_type qualifier

    Reference:
    https://www.insdc.org/submitting-standards/controlled-vocabulary-moltype-qualifier/
    """

    GENOMIC_DNA = "genomic DNA"
    OTHER_DNA = "other DNA"
    UNASSIGNED_DNA = "unassigned DNA"
    GENOMIC_RNA = "genomic RNA"
    MRNA = "mRNA"
    TRNA = "tRNA"
    TRANSCRIBED_RNA = "transcribed RNA"
    VIRAL_CRNA = "viral cRNA"
    OTHER_RNA = "other RNA"

    @classmethod
    def from_molecule(cls: type, molecule: Molecule) -> "INSDCMolType":
        """Return an INSDC moltype value that matches the passed ref-builder molecule."""

        match molecule.type:
            case MolType.DNA:
                return INSDCMolType.GENOMIC_DNA
            case MolType.RNA:
                return INSDCMolType.GENOMIC_RNA
            case MolType.CRNA:
                return INSDCMolType.VIRAL_CRNA
            case MolType.MRNA:
                return INSDCMolType.MRNA
            case MolType.TRNA:
                return INSDCMolType.TRANSCRIBED_RNA
            case _:
                raise ValueError(
                    f"Molecule cannot be matched to INSDCMolType: {molecule.type}"
                )

    def to_molecule(self) -> MolType:
        """Return ref-builder molecule type."""
        if "DNA" in self.value:
            return MolType.DNA

        match self:
            case INSDCMolType.GENOMIC_RNA:
                return MolType.RNA
            case INSDCMolType.VIRAL_CRNA:
                return MolType.CRNA
            case INSDCMolType.MRNA:
                return MolType.MRNA
            case INSDCMolType.TRANSCRIBED_RNA:
                return MolType.TRNA
            case _:
                raise ValueError(
                    f"INSDC molecule type cannot be matched to MolType: {self}"
                )


class INSDCFeatureQualifier(BaseModel):
    """A feature qualifier in a GenBank source table."""

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    name: Annotated[str, Field(validation_alias=FTGenbankLexicon.QUALIFIER_NAME)]
    """The name of the qualifier."""

    value: Annotated[
        str | None, Field(validation_alias=FTGenbankLexicon.QUALIFIER_VALUE)
    ] = None
    """The value of the qualifier. 
    If this field is not present, the qualifier is a flag.
    """


class INSDCFeatureInterval(BaseModel):
    """An interval in a GenBank source table."""

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    start: Annotated[
        int,
        Field(validation_alias=FTGenbankLexicon.INTERVAL_FROM),
    ]
    """The start of the interval."""

    end: Annotated[
        int,
        Field(validation_alias=FTGenbankLexicon.INTERVAL_TO),
    ]
    """The end of the interval."""


class INSDCFeature(BaseModel):
    """An item in a feature table."""

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    key: Annotated[
        str,
        Field(validation_alias=FTGenbankLexicon.FEATURE_KEY),
    ]
    """The name of the feature."""

    intervals: Annotated[
        list[INSDCFeatureInterval],
        Field(validation_alias=FTGenbankLexicon.FEATURE_INTERVALS),
    ]
    """The intervals of the feature."""

    qualifiers: Annotated[
        list[INSDCFeatureQualifier] | None,
        Field(validation_alias=FTGenbankLexicon.FEATURE_QUALIFIERS),
    ] = None
    """The qualifiers of the feature."""


class INSDCFeatureSource(INSDCFeature):
    """Normalized data from a source feature."""

    key: Annotated[
        Literal["source"],
        Field(validation_alias=FTGenbankLexicon.FEATURE_KEY),
    ]

    qualifiers: Annotated[
        list[INSDCFeatureQualifier],
        Field(validation_alias=FTGenbankLexicon.FEATURE_QUALIFIERS),
    ]


class INSDCFeatureTable(RootModel):
    """An unsorted INSDC feature table."""

    root: list[INSDCFeature]

    def __iter__(self) -> Iterator[INSDCFeature]:
        return iter(self.root)
