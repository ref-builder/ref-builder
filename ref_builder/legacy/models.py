from dataclasses import dataclass
from enum import Enum
from typing import Annotated

from pydantic import (
    BaseModel,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)


def transform_to_uppercase(string: str) -> str:
    return string.upper()


class LegacySourceType(str, Enum):
    CLONE = "clone"
    CULTURE = "culture"
    ISOLATE = "isolate"
    GENBANK = "genbank"
    GENOTYPE = "genotype"
    SEROTYPE = "serotype"
    STRAIN = "strain"
    UNKNOWN = "unknown"
    VARIANT = "variant"


@dataclass
class LegacyIsolateSource:
    name: str
    type: LegacySourceType


class LegacySequence(BaseModel):
    id: str = Field(alias="_id")
    accession: Annotated[
        str,
        StringConstraints(
            min_length=4,
            pattern=r"^\w+\.\d+$",
            strip_whitespace=True,
            to_upper=True,
        ),
    ]
    definition: Annotated[str, StringConstraints(min_length=10, strip_whitespace=True)]
    host: Annotated[str, StringConstraints(strip_whitespace=True)]
    segment: str
    sequence: Annotated[
        str,
        StringConstraints(
            min_length=10,
            pattern="^[ATCGRYKMSWBDHVNatcgrykmswbdhvn]+$",
            strip_whitespace=True,
            to_upper=True,
        ),
    ]


class LegacyIsolate(BaseModel):
    id: str
    default: bool
    sequences: Annotated[list[LegacySequence], StringConstraints(min_length=1)]
    source_type: LegacySourceType
    source_name: Annotated[str, StringConstraints(strip_whitespace=True)]

    @model_validator(mode="after")
    def check_source_name(self: "LegacyIsolate") -> "LegacyIsolate":
        """Check if the source name is not empty."""
        if not self.source_name and self.source_type is not LegacySourceType.UNKNOWN:
            raise ValueError(
                "Source name cannot be empty unless source type is unknown"
            )

        return self


class LegacySchemaSegment(BaseModel):
    molecule: str
    name: str
    required: bool


class LegacyOTU(BaseModel):
    id: Annotated[str, Field(alias="_id")]
    abbreviation: Annotated[str, StringConstraints(strip_whitespace=True)]
    isolates: Annotated[list[LegacyIsolate], Field(min_length=1)]
    name: Annotated[str, StringConstraints(min_length=5, strip_whitespace=True)]
    otu_schema: Annotated[
        list[LegacySchemaSegment],
        Field(min_length=1, alias="schema"),
    ]
    taxid: int

    @field_validator("isolates")
    @classmethod
    def check_default_isolate(cls, v):
        """Check if there is only one default isolate."""
        default_isolates = [isolate for isolate in v if isolate.default]

        if len(default_isolates) == 0:
            raise ValueError("At least one isolate must be default")

        if len(default_isolates) > 1:
            raise ValueError("Only one isolate can be default")

        return v

    @field_validator("otu_schema")
    @classmethod
    def check_schema_molecule(cls, otu_schema: list[LegacySchemaSegment]):
        """Check if all segments in the schema have the same molecule."""
        molecules = {segment.molecule for segment in otu_schema}

        if len(molecules) > 1:
            raise ValueError("All segments in a schema must have the same molecule")

        return otu_schema

    @field_validator("otu_schema")
    @classmethod
    def check_schema_names(cls, otu_schema: list[LegacySchemaSegment]):
        """Check if there are duplicate schema segment names."""
        names = {segment.name for segment in otu_schema}

        if len(names) != len(otu_schema):
            raise ValueError("All schema segments must have a unique name")

        return otu_schema

    @model_validator(mode="after")
    def check_schema_and_sequences(self) -> "LegacyOTU":
        """Make sure all sequences in the OTU have a corresponding schema segment."""
        schema_segments = {segment.name for segment in self.otu_schema}

        required_schema_segments = {
            segment.name for segment in self.otu_schema if segment.required
        }

        for isolate in self.isolates:
            for sequence in isolate.sequences:
                if sequence.segment not in schema_segments:
                    raise ValueError(
                        "sequence contains invalid segment name",
                    )

            if required_schema_segments - {
                sequence.segment for sequence in isolate.sequences
            }:
                raise ValueError(
                    "isolate does not contain all required schema segments",
                )

        sequence_segments = {
            sequence.segment
            for isolate in self.isolates
            for sequence in isolate.sequences
        }

        if schema_segments != sequence_segments:
            raise ValueError("All schema segments must have a corresponding sequence")

        return self
