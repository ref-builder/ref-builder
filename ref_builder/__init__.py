"""Virtool CLI for managing references.

TODO: Switch back to Rich.
TODO: Allow exclusion checks at the Repo level.
TODO: Don't use the term "blacklist". Use the word "exclusions" only.
"""
