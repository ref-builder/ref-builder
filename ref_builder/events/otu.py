from pydantic import UUID4

from ref_builder.events.base import EventData, Event, OTUQuery
from ref_builder.models import Molecule
from ref_builder.plan import MonopartitePlan, MultipartitePlan


class CreateOTUData(EventData):
    """The data for a :class:`CreateOTU` event."""

    id: UUID4
    acronym: str
    legacy_id: str | None
    molecule: Molecule
    name: str
    taxid: int
    plan: MonopartitePlan | MultipartitePlan


class CreateOTU(Event):
    """An event that creates a new OTU."""

    data: CreateOTUData
    query: OTUQuery


class ExcludeAccessionData(EventData):
    """The data for a :class:`ExcludeAccession` event."""

    accession: str


class ExcludeAccession(Event):
    """An accession exclusion event.

    This event is emitted when a Genbank accession is not going to be allowed in the
    reference.
    """

    data: ExcludeAccessionData
    query: OTUQuery


class CreatePlanData(EventData):
    """The data for a :class:`CreatePlan` event."""

    plan: MonopartitePlan | MultipartitePlan


class CreatePlan(Event):
    """An event that sets the isolate plan for an OTU."""

    data: CreatePlanData
    query: OTUQuery


class SetReprIsolateData(EventData):
    """The data for a :class:`SetReprIsolate` event."""

    isolate_id: UUID4


class SetReprIsolate(Event):
    """An event that sets the representative isolate for an OTU."""

    data: SetReprIsolateData
    query: OTUQuery
