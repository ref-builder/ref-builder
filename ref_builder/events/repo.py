from pydantic import UUID4

from ref_builder.events.base import EventData, Event, RepoQuery
from ref_builder.resources import RepoSettings


class CreateRepoData(EventData):
    """The data for a :class:`CreateRepo` event."""

    id: UUID4
    data_type: str
    name: str
    organism: str
    settings: RepoSettings


class CreateRepo(Event):
    """An event that creates a new repository.

    This event is always the first event in a repository's event log.
    """

    data: CreateRepoData
    query: RepoQuery
