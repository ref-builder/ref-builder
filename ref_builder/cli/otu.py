import sys
from pathlib import Path
from uuid import UUID

import click
from click import Context
import structlog

from ref_builder.cli.validate import validate_no_duplicate_accessions
from ref_builder.console import print_otu, print_otu_list
from ref_builder.options import ignore_cache_option, path_option
from ref_builder.otu.create import create_otu
from ref_builder.otu.isolate import (
    add_unnamed_isolate,
    add_and_name_isolate,
    add_genbank_isolate,
)
from ref_builder.otu.modify import (
    exclude_accessions_from_otu,
    set_representative_isolate,
    add_segments_to_plan,
    resize_monopartite_plan,
)
from ref_builder.otu.update import auto_update_otu, promote_otu_accessions
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.plan import SegmentRule, SegmentName
from ref_builder.repo import Repo
from ref_builder.utils import IsolateNameType, IsolateName

logger = structlog.get_logger()


@click.group(name="otu")
def otu() -> None:
    """Manage OTUs."""


@otu.command(name="create")
@click.argument("TAXID", type=int)
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--acronym", type=str, default="")
@ignore_cache_option
@path_option
def otu_create(
    accessions_: list[str],
    acronym: str,
    ignore_cache: bool,
    path: Path,
    taxid: int,
) -> None:
    """Create a new OTU.

    OTUs are created from a list of accessions and a taxonomy ID. The associated Genbank
    records are fetched and used to create the first isolate and a plan.
    """
    repo = Repo(path)

    if len(accessions_) != len(set(accessions_)):
        click.echo("Duplicate accessions were provided.", err=True)
        sys.exit(1)

    try:
        create_otu(
            repo,
            taxid,
            accessions_,
            acronym=acronym,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@otu.command(name="get")
@click.argument("IDENTIFIER", type=str)
@path_option
def otu_get(identifier: str, path: Path) -> None:
    """Get an OTU by its unique ID or taxonomy ID."""
    try:
        identifier = int(identifier)
        otu_ = Repo(path).get_otu_by_taxid(identifier)
    except ValueError:
        otu_ = Repo(path).get_otu(UUID(identifier))

    if otu_ is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    print_otu(otu_)


@otu.command(name="list")
@path_option
def otu_list(path: Path) -> None:
    """List all OTUs in the repository."""
    print_otu_list(Repo(path).iter_minimal_otus())


@otu.group(invoke_without_command=True)
@click.argument("TAXID", type=int)
@path_option
@click.pass_context
def update(ctx: Context, path: Path, taxid: int) -> None:
    """Update the specified OTU with new data."""
    repo = Repo(path)

    ctx.ensure_object(dict)
    ctx.obj = {
        "REPO": repo,
        "TAXID": taxid,
    }

    if not repo.get_otu_id_by_taxid(taxid):
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if ctx.invoked_subcommand is None:
        click.echo(ctx.get_help())


@update.command(name="automatic")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_auto_update(ctx: Context, ignore_cache: bool) -> None:
    """Automatically update an OTU with the latest data from NCBI."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        click.echo(f'Run "virtool otu create {taxid} --autofill" instead.')
        sys.exit(1)

    auto_update_otu(repo, otu_, ignore_cache=ignore_cache)


@update.command(name="promote")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_promote_accessions(
    ctx: Context,
    ignore_cache: bool,
) -> None:
    """Promote all RefSeq accessions within this OTU."""
    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(ctx.obj["TAXID"])

    promote_otu_accessions(repo, otu_, ignore_cache)


@update.command(name="isolate")  # type: ignore
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--unnamed",
    is_flag=True,
    default=False,
    help="ignore isolate names in Genbank sources",
)
@click.option(
    "--name",
    type=(IsolateNameType, str),
    help='an overriding name for the isolate, e.g. "isolate ARWV1"',
)
@ignore_cache_option
@click.pass_context
def isolate_create(
    ctx: Context,
    accessions_: list[str],
    ignore_cache: bool,
    name: tuple[IsolateNameType, str] | None,
    unnamed: bool,
) -> None:
    """Create a new isolate using the given accessions."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if unnamed:
        add_unnamed_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )

    if name is not None:
        isolate_name_type, isolate_name_value = name
        isolate_name_ = IsolateName(type=isolate_name_type, value=isolate_name_value)

        try:
            add_and_name_isolate(
                repo,
                otu_,
                accessions_,
                ignore_cache=ignore_cache,
                isolate_name=isolate_name_,
            )
        except RefSeqConflictError as err:
            click.echo(
                f"{err.isolate_name} already exists, but RefSeq items may be "
                "promotable.",
            )
            sys.exit(1)

    try:
        add_genbank_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except RefSeqConflictError as err:
        click.echo(
            f"{err.isolate_name} already exists, but RefSeq items may be promotable,",
        )
        sys.exit(1)

    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@update.command(name="exclude")  # type: ignore
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.pass_context
def accession_exclude(
    ctx: Context,
    accessions_: list[str],
) -> None:
    """Exclude the given accessions from this OTU.

    Any duplicate accessions will be ignored. Only one exclusion per unique accession
    will be made.
    """
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    exclude_accessions_from_otu(repo, otu_, accessions_)


@update.command(name="default")  # type: ignore
@click.argument("ISOLATE_KEY", type=str)
@click.pass_context
def otu_set_representative_isolate(
    ctx: Context,
    isolate_key: str,
) -> None:
    """Update the OTU with a new representative isolate."""
    taxid = ctx.obj["TAXID"]

    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        sys.exit(1)

    try:
        isolate_id = UUID(isolate_key)
    except ValueError:
        parts = isolate_key.split(" ")
        try:
            isolate_name = IsolateName(IsolateNameType(parts[0].lower()), parts[1])
        except ValueError:
            click.echo(f'Error: "{isolate_key}" is not a valid isolate name.', err=True)
            sys.exit(1)

        isolate_id = otu_.get_isolate_id_by_name(isolate_name)

    if isolate_id is None:
        click.echo("Isolate could not be found in this OTU.", err=True)
        sys.exit(1)

    set_representative_isolate(repo, otu_, isolate_id)


@update.group()
@click.pass_context
def plan(ctx: Context) -> None:
    """Add to and replace isolate plans for this OTU."""


@plan.command(name="extend")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_extend_segment_list(
    ctx: Context,
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Add recommended or optional segments to the OTU plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        add_segments_to_plan(
            repo,
            otu_,
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)


@plan.command(name="multipartite")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--name",
    type=(str, str),
    help='a name for the originalsegment, e.g. "RNA A"',
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_resize_monopartite(
    ctx: Context,
    name: tuple[str, str],
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Change a monopartite OTU plan to a multipartite plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    prefix, key = name

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        resize_monopartite_plan(
            repo,
            otu_,
            name=SegmentName(prefix=prefix, key=key),
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
