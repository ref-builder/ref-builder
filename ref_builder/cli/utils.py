import sys
import warnings
from uuid import UUID

import click

from ref_builder.errors import InvalidInputError, PartialIDConflictError
from ref_builder.otu.builders.otu import OTUBuilder
from ref_builder.repo import Repo
from ref_builder.utils import OTUDeletedWarning

pass_repo = click.make_pass_decorator(Repo)

UUID_STRING_LENGTH = 36


def get_otu_from_identifier(repo: Repo, identifier: str) -> OTUBuilder:
    """Return an OTU id from the repo if identifier matches a single OTU.

    The identifier can either be a stringified UUIDv4, a truncated portion
    of a UUID, a NCBI Taxonomy ID or an acronym associated with the OTU.

    Raise a PartialIDConflictError if >1 OTU is found.

    :param repo: the repository to be searched.
    :param identifier: a non-UUID identifier.
        Can be an integer Taxonomy ID, acronym or truncated partial UUID.
    :return: the UUID of the OTU or ``None``
    """
    otu_id = None

    if len(identifier) == UUID_STRING_LENGTH:
        try:
            otu_id = UUID(identifier)
        except ValueError:
            otu_id = None

    elif identifier.isnumeric():
        try:
            taxid = int(identifier)
        except ValueError:
            pass
        else:
            otu_id = repo.get_otu_id_by_taxid(taxid)

    else:
        if (otu_id := repo.get_otu_id_by_acronym(identifier)) is not None:
            return repo.get_otu(otu_id)

        try:
            otu_id = repo.get_otu_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError as e:
            click.echo(
                e,
                err=True,
            )

    if otu_id is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    with warnings.catch_warnings(
        category=OTUDeletedWarning, record=True
    ) as warning_list:
        otu_ = repo.get_otu(otu_id)

    if otu_ is None:
        if warning_list:
            click.echo("OTU has been deleted.", err=True)

        else:
            click.echo("OTU not found.", err=True)

        sys.exit(1)

    return otu_


def get_otu_isolate_ids_from_identifier(repo: Repo, identifier: str) -> (UUID, UUID):
    """Return an isolate ID from the repo if identifier matches a single isolate.

    Handles cases where the isolate cannot be found.
    """
    isolate_id = None

    if len(identifier) == 32:
        try:
            isolate_id = UUID(identifier)
        except ValueError:
            isolate_id = None

    else:
        try:
            isolate_id = repo.get_isolate_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError:
            click.echo(
                "Partial ID segment must be at least 8 characters long.", err=True
            )
            sys.exit(1)

    if isolate_id is not None:
        if (otu_id := repo.get_otu_id_by_isolate_id(isolate_id)) is not None:
            if isolate_id in repo.get_otu(otu_id).isolate_ids:
                return otu_id, isolate_id

    click.echo("Isolate ID could not be found.", err=True)
    sys.exit(1)
