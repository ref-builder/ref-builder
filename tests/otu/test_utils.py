import pytest
from syrupy import SnapshotAssertion
from syrupy.filters import props

from ref_builder.ncbi.client import NCBIClient
from ref_builder.ncbi.models import NCBISourceMolType
from ref_builder.otu.utils import create_plan_from_records
from ref_builder.otu.assign import assign_records_to_segments
from ref_builder.plan import (
    Plan,
)
from ref_builder.repo import Repo
from tests.fixtures.factories import NCBIGenbankFactory, NCBISourceFactory
from tests.fixtures.utils import uuid_matcher


class TestCreatePlanFromRecords:
    """Test `create_plan_from_records` function."""

    def test_monopartite(
        self,
        scratch_ncbi_client: NCBIClient,
        snapshot: SnapshotAssertion,
    ):
        """Test that a monopartite plan is created from a single Genbank record."""
        records = scratch_ncbi_client.fetch_genbank_records(["NC_024301"])

        plan = create_plan_from_records(records, length_tolerance=0.03)

        assert isinstance(plan, Plan)
        assert plan.model_dump() == snapshot(matcher=uuid_matcher)

    def test_multipartite(
        self,
        scratch_ncbi_client: NCBIClient,
        snapshot: SnapshotAssertion,
    ):
        """Test that a multipartite plan is created from multiple Genbank records."""
        records = scratch_ncbi_client.fetch_genbank_records(
            [
                "NC_010314",
                "NC_010315",
                "NC_010316",
                "NC_010317",
                "NC_010318",
                "NC_010319",
            ]
        )

        plan = create_plan_from_records(records, length_tolerance=0.03)

        assert isinstance(plan, Plan)
        assert plan.model_dump() == snapshot(matcher=uuid_matcher)

        segment_names = [str(segment.name) for segment in plan.segments]

        # Make sure the segment names are ordered alphabetically.
        assert segment_names == sorted(segment_names)

    def test_numeric_sorting(self, scratch_ncbi_client: NCBIClient):
        """Test that segment names are sorted numerically."""
        records = scratch_ncbi_client.fetch_genbank_records(
            ["NC_010314", "NC_010315", "NC_010316"]
        )

        for record, segment_name in zip(
            records, ["RNA 3", "RNA 10", "RNA 1"], strict=True
        ):
            record.source.segment = segment_name

        plan = create_plan_from_records(records, length_tolerance=0.03)

        assert [str(segment.name) for segment in plan.segments] == [
            "RNA 1",
            "RNA 3",
            "RNA 10",
        ]


@pytest.mark.parametrize("taxid", [223262, 438782])
class TestAssignRecordsToSegments:
    """Test whether a list of records can be assigned to plan segments."""

    def test_ok(
        self,
        ncbi_genbank_factory: NCBIGenbankFactory,
        ncbi_source_factory: NCBISourceFactory,
        scratch_repo: Repo,
        snapshot: SnapshotAssertion,
        taxid: int,
    ):
        """Test that a list of records generated to match the OTU plan
        can pass assign_records_to_segments().
        """
        otu = scratch_repo.get_otu_by_taxid(taxid)

        records = [
            ncbi_genbank_factory.build(
                source=ncbi_source_factory.build(
                    mol_type=NCBISourceMolType.from_molecule(otu.molecule),
                    segment=str(segment.name),
                    organism=otu.name,
                    taxid=otu.taxid,
                )
            )
            for segment in otu.plan.required_segments
        ]

        assigned_records = assign_records_to_segments(records, otu.plan)

        assert {
            str(otu.plan.get_segment_by_id(segment_id).name): {
                "accession_version": assigned_records[segment_id].accession,
                "source_segment": assigned_records[segment_id].source.segment,
            }
            for segment_id in assigned_records
        } == snapshot(exclude=props("id"))

    def test_names_not_in_plan(
        self,
        ncbi_genbank_factory: NCBIGenbankFactory,
        scratch_repo: Repo,
        snapshot: SnapshotAssertion,
        taxid: int,
    ):
        """Test that a randomly generated list of records raises an appropriate
        ValueError.
        """
        otu = scratch_repo.get_otu_by_taxid(taxid)

        records = ncbi_genbank_factory.batch(len(otu.plan.required_segments))

        with pytest.raises(ValueError, match="Segment names not found in plan:") as e:
            assign_records_to_segments(records, otu.plan)

        assert str(e.value) == snapshot(exclude=props("id"))
