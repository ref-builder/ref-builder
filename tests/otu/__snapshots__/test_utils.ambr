# serializer version: 1
# name: TestAssignRecordsToSegments.test_names_not_in_plan
  'Segment names not found in plan: DNA 1.'
# ---
# name: TestAssignRecordsToSegments.test_ok
  set({
    tuple(
      SegmentName(prefix='DNA', key='A'),
      'NC_907439',
      'DNA A',
    ),
    tuple(
      SegmentName(prefix='DNA', key='B'),
      'K93863',
      'DNA B',
    ),
  })
# ---
# name: TestCreatePlanFromRecords.test_monopartite
  dict({
    'id': UUID,
    'segments': list([
      dict({
        'id': UUID,
        'length': 159,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'B',
          'prefix': 'RNA',
        }),
        'rule': 'required',
      }),
    ]),
  })
# ---
# name: TestCreatePlanFromRecords.test_multipartite
  dict({
    'id': UUID,
    'segments': list([
      dict({
        'id': UUID,
        'length': 1015,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'C',
          'prefix': 'DNA',
        }),
        'rule': 'required',
      }),
      dict({
        'id': UUID,
        'length': 1074,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'M',
          'prefix': 'DNA',
        }),
        'rule': 'required',
      }),
      dict({
        'id': UUID,
        'length': 1090,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'N',
          'prefix': 'DNA',
        }),
        'rule': 'required',
      }),
      dict({
        'id': UUID,
        'length': 1099,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'R',
          'prefix': 'DNA',
        }),
        'rule': 'required',
      }),
      dict({
        'id': UUID,
        'length': 1087,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'S',
          'prefix': 'DNA',
        }),
        'rule': 'required',
      }),
      dict({
        'id': UUID,
        'length': 1057,
        'length_tolerance': 0.03,
        'name': dict({
          'key': 'U3',
          'prefix': 'DNA',
        }),
        'rule': 'required',
      }),
    ]),
  })
# ---
